src/translator/service.cpp

```diff
@@ -5,25 +5,74 @@
 #include <string>
 #include <utility>
 
+inline std::vector<marian::Ptr<const marian::Vocab>>
+loadVocabularies(marian::Ptr<marian::Options> options) {
+  // @TODO: parallelize vocab loading for faster startup
+  auto vfiles = options->get<std::vector<std::string>>("vocabs");
+  // with the current setup, we need at least two vocabs: src and trg
+  ABORT_IF(vfiles.size() < 2, "Insufficient number of vocabularies.");
+  std::vector<marian::Ptr<marian::Vocab const>> vocabs(vfiles.size());
+  std::unordered_map<std::string, marian::Ptr<marian::Vocab>> vmap;
+  for (size_t i = 0; i < vocabs.size(); ++i) {
+    auto m =
+        vmap.emplace(std::make_pair(vfiles[i], marian::Ptr<marian::Vocab>()));
+    if (m.second) { // new: load the vocab
+      m.first->second = marian::New<marian::Vocab>(options, i);
+      m.first->second->load(vfiles[i]);
+    }
+    vocabs[i] = m.first->second;
+  }
+  return vocabs;
+}
+
 namespace marian {
 namespace bergamot {
 
-Service::Service(Ptr<Options> options, const void * model_memory)
-    : ServiceBase(options), numWorkers_(options->get<int>("cpu-threads")),
-      pcqueue_(numWorkers_), model_memory_{model_memory} {
+Service::Service(Ptr<Options> options, const void *model_memory)
+    : requestId_(0), vocabs_(std::move(loadVocabularies(options))),
+      text_processor_(vocabs_, options), batcher_(options),
+      numWorkers_(options->get<int>("cpu-threads")), model_memory_(model_memory)
+#ifndef WASM_HIDE_THREADS
+      ,
+      pcqueue_(numWorkers_)
```

Comment:
`numWorkers_ = 0` is illegal for PCQueue. This is a bug. Could hotfix with a `pcqueue(safetyGuard(numWorkers_))` to have at least one element.

Legal init is no pcqueue in class from `cpu-threads=0`, also no vector<std::threads>.

Reply:
Yeah a pcqueue with 0 size doesn't make much sense.  
`pcqueue_(std::max<std::size_t>(1, numWorkers_))` + comment that you won't use it anyway?

Reply:
My problem is that this is building and running (passing tests) on var, while it shouldn't. The source has also arbitrarily failed on me before. That's partly why I so badly want two classes doing two things (single threaded no pcqueue, no std::thread, and multithreaded with pcqueue and std::thread). unitialized `std::vector<std::thread>` is also a problem. 

Everywhere I read about this, conditional compilation is not favourable here (w.r.t having members with legal inits). I'm caving to complaints from everywhere, my gut feeling is we should separate single and multithreaded implementations binding them in a common interface.

Reply:
What do you mean it's unitialized?  It is empty.

Reply:
Sure hope so, read in forums it is meant by standard to be so as well. But destructor gave segfaults with range iterations - `for(auto &worker: workers)` having illegal values when I tried to inspect https://github.com/browsermt/bergamot-translator/issues/41 running through debugger. I had to clear that to get to the bottom of #41 (which I did by separating single-threaded and multithreaded implementation which felt like the saner thing to do for future development), but by then segfaults went away.

In this PR, keeping that in mind, I have changed that part to use `numWorkers_` instead which has legal values.

Reply:
std::vector is ALWAYS initialized to empty by the default constructor.  And a for loop over an empty vector should not produce "illegal values".